CaloMC/src/CaloNoiseSimGenerator.cc

```diff
@@ -0,0 +1,220 @@
+     diagLevel_     (config.diagLevel())
+   {}       
+
+   
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::initialize(const CaloWFExtractor& wfExtractor)
+   {
+       pulseShape_.buildShapes();
+       
+       generateWF(waveform_);
+       if (enableAR_) initAR();
+       generateFragments(wfExtractor);
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::generateWF(std::vector<double>& wfVector)
+   {     
+       ConditionsHandle<CalorimeterCalibrations> calorimeterCalibrations("ignored");
+       double scaleFactor = calorimeterCalibrations->MeV2ADC(iRO_)/calorimeterCalibrations->peMeV(iRO_);
+
+       std::fill(wfVector.begin(),wfVector.end(),0);
+
+       const auto&      pulse         = pulseShape_.digitizedPulse(0.0);
+       const unsigned   pulseSize     = pulse.size();
+       const unsigned   bufferSize    = int(0.75*pulseSize);
+       const unsigned   noiseSize     = wfVector.size();
+       const double     totalTime     = (noiseSize+bufferSize)*digiSampling_;
+       const int        noiseLevelPE  = int(totalTime*noiseRinDark_);
+
+       //Generate the radiation induced noise (RIN)
+       const int nPh = randPoisson_(noiseLevelPE);
+       for (int i=0;i<nPh;++i)
+       {
+	   double t0 = randFlat_.fire(0.0,totalTime);       
+	   const auto& wf = pulseShape_.digitizedPulse(t0);
+
+	   int i0 = int(t0/digiSampling_) - bufferSize;
+	   int l0 = (i0<0) ? -i0 : 0;      
+	   int l1 = std::min(pulseSize,noiseSize-i0);
+	   for (int l=l0;l<l1;++l) wfVector[i0+l] += wf[l]*scaleFactor;
+       }
+
+       //add electronics noise        
+       double noiseADC = noiseElec_*digiSampling_*scaleFactor;
+       for (auto& val : wfVector) val += randGauss_.fire(0.0,noiseADC);
+
+       //estimate pedestal for this waveform - set it to theoretical value for the time being
+       pedestal_ = int(noiseRinDark_*digiSampling_*std::accumulate(pulse.begin(),pulse.end(),0.0)*scaleFactor);
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::initAR()
+   {
+       std::vector<double> wf(waveform_);
+       for (auto& val : wf) val -= pedestal_;
+
+       ARFitter_.setWaveform(wf);
+       ARFitter_.fitARCoeff();
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::generateFragments(const CaloWFExtractor& wfExtractor)
+   {
+       unsigned nwf(0), nfound(0), length(255);
+       for (;nwf<nMaxFragment_;++nwf)
+       {
+          std::vector<double> temp(length,0.0);
+          addFullNoise(temp,false);
+
+          std::vector<int> wf;
+          for (const auto& val : temp) wf.emplace_back(val - pedestal_);    
```

Comment:
Please do a reserve before filling the vector.  If compilers have got smarter and know to do the reserve on it's own, let me know.   Please check other parts of the code for the same issue - I noticed a few more spots.